Attested TLS APIs proposal (#4138)
openenclave/openenclave: docs/DesignDocs/AttestationAPIsWithFlexibleClaims.md

- Limited support for run-time claims
  The notion of run-time claims is roughly equivalent to the custom claims
  in the OE attestation plugin framework. The `oe_get_attestation_certificate_with_evidence_v2`
  API does not allow developers to provide customized custom claims as input.
  Instead, the API internally takes the input public key buffer as the custom claim.
  The `oe_verify_attestation_certificate_with_evidence_v2` then extracts the custom
  claim and verifies its value against the public key.

Proposal
--------

This document proposes new APIs,
`oe_get_attestation_certificate_with_evidence_v3` and
`oe_verify_attestation_certificate_with_evidence_v3`, that
address the abovementioned limitations. Note that the certificate
generated by `oe_get_attestation_certificate_with_evidence_v3`
does not provide full backward compatibility to older version
of verification APIs (e.g., `oe_verify_attestation_certificate_with_evidence_v2`
and `oe_verify_attestation_certificate`).

- oe_get_attestation_certificate_with_evidence_v3
  ```c
  /**
   * oe_get_attestation_certificate_with_evidence_v3
   *
   * Similar to oe_get_attestation_certificate_with_evidence_2, this function

>    * generates a self-signed x.509 certificate with embedded evidence generated by

Reviewer: ```suggestion
   * generates a self-signed X.509 certificate with embedded evidence generated by
```

   * an attester plugin for the enclave, but it also allows a user to pass in
   * runtime and inittime claims.
   *
   * @experimental
   *
   * @param[in] format_id The format id of the evidence to be generated.
   *
   * @param[in] subject_name a string containing an X.509 distinguished
   * name (DN) for customizing the generated certificate. This name is also used

>    * as the issuer name because this is a self-signed certificate

Reviewer: ```suggestion
   * as the issuer name because this is a self-signed certificate.
```

>    * See RFC5280 (https://tools.ietf.org/html/rfc5280) for details

Reviewer: ```suggestion
   * See RFC5280 (https://tools.ietf.org/html/rfc5280) for details.
```

   * Example value "CN=Open Enclave SDK,O=OESDK TLS,C=US"
   *

>    * @param[in] private_key a private key used to sign this certificate

Reviewer: nit: half the lines have the param description start with a capital letter, and have with a lower case letter.
Similarly, half end in a period and half don't.  Be consistent.

   * @param[in] private_key_size The size of the private_key buffer
   * @param[in] public_key a public key used as the certificate's subject key
   * @param[in] public_key_size The size of the public_key buffer.
   * @param[in] inittime_claims an optional flat buffer representing the inittime
   * claims.
   * @param[in] inittime_claims_size The size of the inittime_claims in bytes.
   * @param[in] runtime_claims an optional flat buffer representing the runtime
   * claims.
   * @param[in] runtime_claims_size The size of the runtime_claims in bytes.
   * @param[in] optional_parameters The optional format-specific input parameters.
   * @param[in] optional_parameters_size The size of optional_parameters in bytes.
   *
   * @param[out] output_certificate a pointer to buffer pointer
   * @param[out] output_certificate_size size of the buffer above
   *
   * @return OE_OK on success
   */
  oe_result_t oe_get_attestation_certificate_with_evidence_v3(

>       const oe_uuid_t* format_id,

Reviewer: missing `flags` param that exists for oe_get_evidence().
I will also observe that there is no `endorsements_buffer` output like get_evidence has, why is that?

Author: The evidence embedded in the certificate uses the  `OE_EVIDENCE_FLAGS_EMBED_FORMAT_ID` flag by default, which is the contract between the get and verify certificate APIs.

I'm not sure why endorsements is not supported (currently the get API passes endorsements  as `NULL`  to `oe_get_evidence`). I'm also unclear about the use case of supporting endorsements in these APIs.
@yentsanglee can you comment on this?

Reviewer 2: endorsements_buffer is a special contract between evidence generating and verifying. Verifier can use the same endorsement retrieved at the time of evidence generation to verify the evidence. This is useful for verifying possibly stale evidence. I don't see a use case for certificate.

Reviewer: Use case is the same as for evidence.

      const unsigned char* subject_name,
      uint8_t* private_key,
      size_t private_key_size,
      uint8_t* public_key,
      size_t public_key_size,
      uint8_t* runtime_claims,
      size_t runtime_claims_size,
      uint8_t* inittime_claims,

>       size_t inittime_claims_size,

Reviewer: InitTimeConfigurationInterface.md has:
```
    const void* runtime_custom_claims_buffer,
    size_t runtime_custom_claims_buffer_size,
    const oe_inittime_custom_claim_buffer_t* inittime_custom_claims_buffer,
    size_t inittime_custom_claims_buffer_size,
```
Be consistent.

      const void* optional_parameters,
      size_t optional_parameters_size,
      uint8_t** output_certificate,
      size_t* output_certificate_size);

  ```
  The API introduces four arguments that allow developers to optionally
  pass in run- and init-time claims.
  - `runtime_claims`
  - `runtime_claims_size`
  - `inittime_claims`
  - `inittime_claims_size`

- oe_verify_attestation_certificate_with_evidence_v3
  ```c
  /**
   * oe_verify_attestation_certificate_with_evidence_v3
   *
   * This function is similar to
   * oe_verify_attestation_certificate_with_evidence_v2 that has the same
   * prototype but with the ability to verify a ceritifacte generated by
   * oe_get_attestation_certificate_with_evidence_v3.
   *
   * @experimental
   *
   * @param[in] certificate_in_der A pointer to buffer holding certificate

>    * contents in DER format.

Reviewer: question: I know this is experimental, but do we really want separate APIs for certificates vs (say) JWTs vs CWTs?
oe_verify_evidence in contrast has a format_id to allow flexibility of format.   I might ask the same about the format of the endorsements_buffer.

Author: This sounds like supporting a format of "certificate with embedded evidence" for the endorsement. In this case, the `evidence_buffer` can be NULL.

```
        result = oe_verify_evidence(
            NULL,
            NULL,
            0,
            endorsements_buffer, /* include the certificate with the embedded evidence */
            endorsements_buffer_size,
            policies,
            policies_size,
            claims,
            claims_length);
``` 

But this requires significant changes to the  `oe_get_evidence` API to support this proposal unless we still keep the `oe_get_attestation_certificate_with_evidence_v3` but drop `oe_verify_attestation_certificate_with_evidence_v3`.

   * @param[in] certificate_in_der_size Size of certificate buffer above.
   * @param[in] endorsements_buffer A pointer to buffer holding endorsements.
   * @param[in] endorsements_buffer_size Size of the endorsements buffer.
   * @param[in] policies A pointer to buffer holding policies.
   * @param[in] policies_size Size of the policies buffer.
   * @param[out] claims A double-pointer to buffer holding claims.
   * @param[out] claims_length Size of the claims buffer if not NULL.
   * @retval OE_OK Successful validation.
   * @retval OE_VERIFY_FAILED Quote failure.
   * @retval OE_INVALID_PARAMETER One or more invalid parameters.
   * @retval OE_FAILURE General failure.
   * @retval Other appropriate error code.
   */
  oe_result_t oe_verify_attestation_certificate_with_evidence_v3(
      uint8_t* certificate_in_der,
      size_t certificate_in_der_size,
      uint8_t* endorsements_buffer,
      size_t endorsements_buffer_size,

>       oe_policy_t* policies,

Reviewer: oe_verify_evidence() makes all the [in] params be const, this should too

   * @param[in] policies A pointer to buffer holding policies.
   * @param[in] policies_size The size of the buffer above.
   * @param[out] claims A double-pointer to buffer holding claims.
   * @param[out] claims_length The size of the buffer above.
   * @retval OE_OK Successful validation.
   * @retval OE_VERIFY_FAILED Quote failure.
   * @retval OE_INVALID_PARAMETER One or more invalid parameters.
   * @retval OE_FAILURE General failure.
   * @retval Other appropriate error code.
   */
  oe_result_t oe_verify_attestation_certificate_with_evidence_v3(
      const uint8_t* certificate_in_der,
      size_t certificate_in_der_size,
      const uint8_t* endorsements_buffer,
      size_t endorsements_buffer_size,
      const oe_policy_t* policies,
      size_t policies_size,
      oe_claim_t** claims,
      size_t* claims_length);
  ```
  The API does not introduce changes to the function prototype. The purpose of
  the API is to pair with `oe_get_attestation_certificate_with_evidence_v3`;
  i.e., only the API is able to verify the certificated generated via
  `oe_get_attestation_certificate_with_evidence_v3` with either `runtime_custom_claims_buffer`

>   or `inittime_custom_claims_buffer` supplied.

Reviewer 3: Why can't we have both init claims and custom claims? And how do we verify runtime claims?

Author: Removed the word `either` that causes confusion.

The current design is that it's up to the verifier to implement the verification logic against the runtime claims output by the the `oe_verify_attestation_certificate_with_evidence_v3`.

+   const uint8_t* inittime_custom_claims_buffer,
+   size_t inittime_custom_claims_buffer_size,
    const void* optional_parameters,
    size_t optional_parameters_size,
    uint8_t** output_certificate,
    size_t* output_certificate_size)
{
    oe_result_t result = OE_FAILURE;
    uint8_t* evidence_buffer = NULL;
    size_t evidence_buffer_size = 0;
+   uint8_t* custom_claims_buffer = NULL;
+   size_t custom_claims_buffer_size = 0;
+   uint8_t* evidence_with_inittime_custom_claims = NULL;
+   size_t evidence_with_inittime_custom_claims_size = 0;

    OE_TRACE_VERBOSE("Calling oe_get_attestation_certificate_with_evidence_v3");
    OE_TRACE_VERBOSE(
        "generate evidence with hash from public_key_size=%d public_key key "
        "=\n[%s]\n",
        public_key_size,
        public_key);

+   custom_claims_buffer_size = public_key_size + runtime_custom_claims_buffer_size;
+   custom_claims_buffer = oe_malloc(custom_claims_buffer_size);
+   if (!custom_claims_buffer)
+       OE_RAISE(OE_OUT_OF_MEMORY);

+   memcpy(custom_claims_buffer, public_key, public_key_size);
+   memcpy(
+       custom_claims_buffer + public_key_size,
+       runtime_custom_claims_buffer,

> +       runtime_custom_claims_buffer_size);

Reviewer 3: The public key could have variable size. Once you copied it into the `custom_claims_buffer`, how do we extract it out?

Author: The design assumes the public_key_size is provided by the `oe_get_attestation_certificate_with_evidence_v3` API. The certificate generated by the API should include the public key with the given same size. The `oe_verify_attestation_certificate_with_evidence_v3` firstly gets the public key from the certificate and then extracts the runtime claim from the custom_claims_buffer.

    result = oe_get_evidence(
        format_id,
        OE_EVIDENCE_FLAGS_EMBED_FORMAT_ID,
-       public_key,
-       public_key_size,
+       custom_claims_buffer,
+       custom_claims_buffer_size,
        optional_parameters,
        optional_parameters_size,
        &evidence_buffer,
        &evidence_buffer_size,
        NULL,
        0);
    OE_CHECK_MSG(
        result, "oe_get_evidence failed with %s\n", oe_result_str(result));

+   evidence_with_inittime_custom_claims_size =
+       evidence_buffer_size + inittime_custom_claims_buffer_size;

> +   evidence_with_inittime_custom_claims =

Reviewer 3: Same question about the init time claims. How can we extract it without the size info?

Author: The design is based on the contract between `oe_get_attestation_certificate_with_evidence_v3` and `oe_verify_attestation_certificate_with_evidence_v3` APIs. Both APIs assume the evidence with the `OE_EVIDENCE_FLAGS_EMBED_FORMAT_ID` format instead of raw report. The evidence with this format comes with a header that includes the size of the evidence.

+       oe_malloc(evidence_with_inittime_custom_claims_size);
+   if (!evidence_with_inittime_custom_claims)
+       OE_RAISE(OE_OUT_OF_MEMORY);

+   memcpy(
+       evidence_with_inittime_custom_claims, evidence_buffer, evidence_buffer_size);
+   memcpy(
+       evidence_with_inittime_custom_claims + evidence_buffer_size,
+       inittime_custom_claims_buffer,
+       inittime_custom_claims_buffer_size);

    result = generate_x509_self_signed_certificate(
        oid_oe_evidence,

>         sizeof(oid_oe_evidence),

Reviewer 3: Watch out for the length limit of the generated certificate. I think we already went to mbedtls once to ask them to increase the length limit with the self signed certificate generated by OE. Now adding inittime and runtime claims definitely will increase the length even further.

Attestation APIs with Init- and Run-Time Claims Support
==========================

Motivation
------------

The Open Enclave SDK currently supports attestation APIs,
`oe_get_attestation_certificate_with_evidence_v2` and
`oe_verify_attestation_certificate_with_evidence_v2`, that
allow developers to obtain a self-signed certificate with
the enclave evidence embedded and to verify the certificate,
respectively. The latter API also outputs an array of claims,
allowing the developers to implement additional verification
logic. However, these APIs have following limitations:

> - Lack of support of init-time custom claims

Reviewer 2: Init-time claims can be entered at the enclave creation time `oe_create_enclave` and stays the same for the life time of the enclave. It is supported. No need to add extra parameters for it here.

Author: CONFIG_DATA (or CONFIG_ID) is used to verify the integrity of the init-time claims. See example in
https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/InitTimeConfigurationInterface.md#plugin-api-change-with-explicit-init-time-custom-claim-buffer-support

```
    /// Verifier Plugin
    ///
    oe_result_t oe_verify_evidence(
    const oe_uuid_t* format_id,
    const uint8_t* evidence_buffer,
    size_t evidence_buffer_size,
    const uint8_t* endorsements_buffer,
    size_t endorsements_buffer_size,
    const oe_policy_t* policies,
    size_t policies_size,
    oe_claim_t** claims,
    size_t* claims_length)
    {
        ...
        //extract the init_time_claims_buffer and verify it
        inittime_claims_buffer = _find_init_time_claim_buffer(evidence_buffer,
                                   evidence_buffer_size);
        inittime_claims_buffer_size = _find_init_time_claim_buffer_size(
                                        evidence_buffer, evidence_buffer_size);
        config_id = _find_config_id(evidence_buffer, evidence_buffer_size);
        if (inittime_claim_buffer.integrity_algorithm_id ! = 0 )
            return error;
        if (SHA256(inittime_claim_buffer.buffer, inittime_claims_buffer_size)
              != config_id)
            return error;
        //output init_time claims
        ...
    }

```

  According to [Open Enclave Init-time Configuration Interface](InitTimeConfigurationInterface.md),
  the support of init-time custom claims is desirable for the incoming SGX2
  features (i.e., `CONFIGID`). The `oe_get_attestation_certificate_with_evidence_v2`
  API does not allow developers to provide an init-time custom claim buffer as input.

- Limited support for run-time custom claims

  The notion of run-time custom claims is equivalent to the custom claims
  in the OE attestation plugin framework. The `oe_get_attestation_certificate_with_evidence_v2`
  API does not allow developers to provide custom claims as input.
  Instead, the API internally takes the input public key buffer as the custom claim by default.
  The `oe_verify_attestation_certificate_with_evidence_v2` then extracts the custom
  claim and verifies its value against the public key.

Specification
--------

This document proposes new APIs,
`oe_get_attestation_certificate_with_evidence_v3` and
`oe_verify_attestation_certificate_with_evidence_v3`, that
address the abovementioned limitations. Note that the certificate
generated by `oe_get_attestation_certificate_with_evidence_v3`

> does not provide full backward compatibility to older version

Reviewer: ```suggestion
does not provide full backward compatibility to older versions
```

of verification APIs (e.g., `oe_verify_attestation_certificate_with_evidence_v2`
and `oe_verify_attestation_certificate`).

- oe_get_attestation_certificate_with_evidence_v3
  ```c
  /**
   * oe_get_attestation_certificate_with_evidence_v3
   *
   * Similar to oe_get_attestation_certificate_with_evidence_2, this function
   * generates a self-signed X.509 certificate with embedded evidence generated by

>    * an attester plugin for the enclave. In addition, it allows a user to pass in

Reviewer: ```suggestion
   * an attester plugin for the enclave. In addition, it allows an app to pass in
```

   * as the issuer name because this is a self-signed certificate.
   * See RFC5280 (https://tools.ietf.org/html/rfc5280) for details.
   * Example value "CN=Open Enclave SDK,O=OESDK TLS,C=US"
   * @param[in] private_key A private key used to sign this certificate.
   * @param[in] private_key_size The size of the buffer above.
   * @param[in] public_key A public key used as the certificate's subject key.
   * @param[in] public_key_size The size of the buffer above.
   * @param[in] inittime_custom_claims_buffer An optional flat buffer representing the init-time
   * custom claims.
   * @param[in] inittime_custom_claims_buffer_size The size of the buffer above in bytes.
   * @param[in] runtime_custom_claims_buffer An optional flat buffer representing the run-time
   * cutom claims.
   * @param[in] runtime_custom_claims_buffer_size The size of the buffer above in bytes.
   * @param[in] optional_parameters The optional format-specific input parameters.
   * @param[in] optional_parameters_size The size of buffer above in bytes.
   * @param[out] output_certificate A pointer to buffer pointer.
   * @param[out] output_certificate_size The size of the buffer above.
   *

>    * @return OE_OK on success

Reviewer: ```suggestion
   * @retval OE_OK The operation was successful.
```
Use retval for discrete return values (like oe_result_t)

   */
  oe_result_t oe_get_attestation_certificate_with_evidence_v3(
      const oe_uuid_t* format_id,
      const unsigned char* subject_name,
      const uint8_t* private_key,
      size_t private_key_size,
      const uint8_t* public_key,
      size_t public_key_size,
      const uint8_t* runtime_custom_claims_buffer,
      size_t runtime_custom_claims_buffer_size,
      const uint8_t* inittime_custom_claims_buffer,
      size_t inittime_custom_claims_buffer_size,
      const void* optional_parameters,
      size_t optional_parameters_size,
      uint8_t** output_certificate,
      size_t* output_certificate_size);

  ```
  The API introduces four arguments that allow developers to optionally
  pass in run- and init-time claims.
  - `runtime_custom_claims_buffer`
  - `runtime_custom_claims_buffer_size`
  - `inittime_custom_claims_buffer`
  - `inittime_custom_claims_buffer_size`

- oe_verify_attestation_certificate_with_evidence_v3
  ```c
  /**
   * oe_verify_attestation_certificate_with_evidence_v3
   *
   * This function is similar to
   * oe_verify_attestation_certificate_with_evidence_v2 that has the same

>    * prototype but with the ability to verify a ceritifacte generated by

Reviewer: ```suggestion
   * prototype but with the ability to verify a certificate generated by
```

   * oe_get_attestation_certificate_with_evidence_v3.
   *
   * @experimental
   *
   * @param[in] certificate_in_der A pointer to buffer holding certificate
   * contents in DER format.
   * @param[in] certificate_in_der_size The size of the buffer above.

>    * @param[in] endorsements_buffer A pointer to buffer holding endorsements.

Reviewer 2: The `endorsements_buffer` should either exist in both *get* and *verify* or none. Otherwise, where do you get the endorsements buffer from?

Author: This is inherent from oe_verify_attestation_certificate_with_evidence_v2, so this seems to be a problem for existing APIs.

https://github.com/openenclave/openenclave/blob/08ebe60c1d2e3ea24ac634c673a420296bff3352/include/openenclave/attestation/verifier.h#L249

  i.e., only the API is able to verify the certificated generated via
  `oe_get_attestation_certificate_with_evidence_v3` with `runtime_custom_claims_buffer`
  or `inittime_custom_claims_buffer` supplied.

User Experience
--------

The following example demonstrates the use of proposed APIs.

```c
static const uint8_t inittime_custom_claims[] =
  { 0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
    0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f };

static const uint8_t runtime_custom_claims[] =
 { 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x66, 0x77,
   0x88, 0x99, 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff };

...

// Attester
uint8_t* output_certificate = nullptr;
size_t output_certificate_size = 0;

...

result = oe_get_attestation_certificate_with_evidence_v3(
     &format,
     (const unsigned char*)"CN=Open Enclave SDK,O=OESDK TLS,C=US",
     private_key,
     private_key_size,
     public_key,
     public_key_size,
     (uint8_t*)runtime_custom_claims,
     sizeof(runtime_custom_claims),
     (uint8_t*)inittime_custom_claims,
     sizeof(inittime_custom_claims),
     NULL,
     0,
     &output_certificate,
     &output_certificate_size);
OE_CHECK(result);

...

> Send output_certificate to verifier

Reviewer: I'm having a hard time understanding the use case here.  I'd expect the certificate to go to the relying party, and only the evidence to go to the verifier.

Author: This is a simplified end-to-end example (i.e., the model used in the attested-tls sample)

Reviewer: The attested tls sample has a tls session between an attester and a relying party

Author: I cannot find that there is a relying party involves in the attested tls sample. Can you point it out?

...

// Verifier (receive input_certificate)
oe_claim_t* claims = nullptr;
size_t claims_length = 0;

result = oe_verify_attestation_certificate_with_evidence_v3(
     input_certificate,
     input_certificate_size,
     nullptr,
     0,
     nullptr,
     0,
     &claims,
     &claims_length);
OE_CHECK(result);

> // Implement veritication logic against claims

Reviewer: ```suggestion
// Implement verification logic against claims
```

    oe_claim_t** claims,
    size_t* claims_length)
{
    oe_result_t result = OE_FAILURE;
    oe_cert_t cert = {0};
    uint8_t* report = NULL;
    size_t report_size = 0;
    oe_attestation_header_t* header = NULL;
    uint8_t* public_key_buffer = NULL;
    size_t public_key_buffer_size = KEY_BUFF_SIZE;

    const char* oid_array[] = {
        oid_oe_report, oid_new_oe_report, oid_oe_evidence, oid_new_oe_evidence};
    size_t oid_array_index = 0;
    size_t oid_array_count = OE_COUNTOF(oid_array);

    public_key_buffer = (uint8_t*)oe_malloc(KEY_BUFF_SIZE);
    if (!public_key_buffer)
        OE_RAISE(OE_OUT_OF_MEMORY);

    result =
        oe_cert_read_der(&cert, certificate_in_der, certificate_in_der_size);
    OE_CHECK_MSG(result, "certificate_in_der_size=%d", certificate_in_der_size);

    // validate the certificate signature
    result = oe_cert_verify(&cert, NULL, NULL, 0);
    OE_CHECK_MSG(
        result,
        "oe_cert_verify failed with error = %s\n",
        oe_result_str(result));

    //------------------------------------------------------------------------
    // Validate the report's trustworthiness
    // Verify the remote report to ensure its authenticity.

>     // set enclave to NULL because we are dealing only with remote report now

Reviewer: Why does this use the word "report"?   That sounds obsolete.

    //------------------------------------------------------------------------

    // determine the size of the extension
    while (oid_array_index < oid_array_count)
    {
        if (oe_cert_find_extension(
                &cert,
                (const char*)oid_array[oid_array_index],
                NULL,
                &report_size) == OE_BUFFER_TOO_SMALL)
        {
            report = (uint8_t*)oe_malloc(report_size);
            if (!report)
                OE_RAISE(OE_OUT_OF_MEMORY);

            OE_CHECK(oe_cert_find_extension(
                &cert,
                (const char*)oid_array[oid_array_index],
                report,
                &report_size));

            break;
        }

        oid_array_index++;
    }

    // if there is no match
    if (oid_array_index == oid_array_count)
        OE_RAISE_MSG(
            OE_FAILURE, "No expected certificate extension matched", NULL);

    // find the extension

>     OE_TRACE_VERBOSE("extract_x509_report_extension() succeeded");

Reviewer: oe_cert_find_extension?

            policies,
            policies_size,
            claims,
            claims_length);

+       if (report_size > evidence_size)
+       {
+           size_t inittime_custom_claims_size = report_size - evidence_size;
+           uint8_t* inittime_custom_claims = report + evidence_size;
+           _add_claim(
+               claims,
+               claims_length,
+               OE_CLAIM_INITTIME_CUSTOM_CLAIMS_BUFFER,
+               sizeof(OE_CLAIM_INITTIME_CUSTOM_CLAIMS_BUFFER),
+               inittime_custom_claims,
+               inittime_custom_claims_size);
+       }
    }
    else // oid_oe_report or oid_new_oe_report
    {
        result = oe_verify_evidence(
            // The format ID is OE_FORMAT_UUID_LEGACY_REPORT_REMOTE for all OE
            // reports for remote attestation.
            &_uuid_legacy_report_remote,
            report,
            report_size,
            endorsements_buffer,
            endorsements_buffer_size,
            policies,
            policies_size,
            claims,
            claims_length);
    }

    OE_CHECK(result);
    OE_TRACE_VERBOSE("quote validation succeeded");

    // verify report data: hash(public key)

>     // extract public key from the cert

Reviewer: ```suggestion
    // extract public key from the certificate
```

    return result;
}
```

Discussion
--------

- Separation of custom claims and run-time custom claims.

  The current design introduces a separated `OE_CLAIM_RUNTIME_CUSTOM_CLAIMS_BUFFER`
  in addition to `OE_CLAIM_CUSTOM_CLAIMS_BUFFER`. Doing so allows for keeping
  the existing behavior of verifying public and providing more flexible use
  of run-time custom claims. The downside is confusing the use of custom claims.
  An alternative is
  - Rename the `OE_CLAIM_CUSTOM_CLAIMS_BUFFER` as `OE_CLAIM_RUNTIME_CUSTOM_CLAIMS_BUFFER`
  - Introduce `OE_CLAIM_RUNTIME_CUSTOM_CLAIMS_PUBLIC_KEY` (or simply `OE_CLAIM_CUSTOM_CLAIMS_PUBLIC_KEY`)

  This topic is open for discussion.

- Verification of init-time custom claims.

  The current design does not implement the logic of verifying init-time custom claims.
  This decision is letting the developers implement such logic. 
  Alternatively, the implementation can enforce the check of hash(`inittime_custom_claims`) against
  `CONFIGID` by default. This topic is also open for discussion.

- Lack of integrity protection with init-time custom claims.

  Based on the current design, the run-time custom claims are integrity-protected (i.e.,

>   its hash is part of the SGX report) while init-time custom claims rely on the

Reviewer: Why is this specific to "SGX reports"?   This design doc should be support multiple evidence formats.

another party verifies the enclave evidence from the certificate based on the
`oe_verify_attestation_certificate_with_evidence_v2` API
(see the [attested TLS sample](/samples/attested_tls) for more detail).
However, these APIs have the following limitations:

- Fail to support general use cases

  A typical attestation procedure includes three roles: an attester, a relying party,
  and a verifier. Two general models are as follows.
  - Passport Model

    The attester generates the evidence and asks the verifier to assess the
    trustworthiness of the evidence. Then, the attester sends the attestation results
    to the relying party. Finally, the relying party validates the attestation results
    based on customized policies.

  - Background-Check Model

    The attester generates the evidence and directly sends the evidence to the
    relying party. The relying party then asks the verifier to assess the trustworthiness
    of the evidence and then validates the attestation results based on customized policies.

  The end-to-end use case supported by the APIs, however, assumes that the party
  that sends the certificate is the attester and the receiver of the certificate plays

>   the roles of both the verifier and the relying party. More specifically, the

Reviewer 4: The existing API has been designed to specifically support the background check mode, with a combined verifier + relying-party entity. The design in this document feels like a patch to stretch the existing API to support the new usages (especially the passport mode). I think it would be a better approach to start a with a top-down design to build a layer on top of the existing API. For passport mode support, I'd like to see a detailed description of the usages, the cert + attestation result design, and the end-to-end flows between the three parties (attester, verifier, relying party).

Author: I agree to have a different APIs other than extending the existing APIs. For example, simply `oe_get_attestation_certificate` and `oe_parse_attestation_certificate`. Please refer the user experience part for the example of work flow.

The biggest concern I have is how much backward compatibility the new APIs need to provide. Also, given that we have the `oe_create_attestation_ceritifcate` API already, we would need a better naming for the new get API.

Reviewer 4: Note: this PR could be seen as an effort to design a higher-level API to is more aligned to Dave's document [Proposed Attestation API Requirements](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/AttestationApiRequirements.md). The existing API can be seen as a lower-level one to build on top.

Reviewer 2: First generation - `oe_generate_attestation_certificate` embeds OE_REPORT in OID 1.3.6.1.4.1.311.105.1
Second generation - `oe_get_attestation_certificate_with_evidence*` embeds OE_EVIDENCE in OID 1.3.6.1.4.1.311.105.2

Report and evidence have subtle difference.

We might consider this new thing embedded in certificate an extension of evidence and give it a new OID (e.g. 1.3.6.1.4.1.311.105.2.1). This will guarantee backward compatibility. We just have to be careful picking OID to avoid collision or misinterpretation.

   *
   * @experimental
   *
   * @param[in] format_id The format id of the evidence to be generated.
   * @param[in] subject_name A string containing an X.509 distinguished
   * name (DN) for customizing the generated certificate. This name is also used
   * as the issuer name because this is a self-signed certificate.
   * See RFC5280 (https://tools.ietf.org/html/rfc5280) for details.
   * Example value "CN=Open Enclave SDK,O=OESDK TLS,C=US"
   * @param[in] private_key A private key used to sign this certificate.
   * @param[in] private_key_size The size of the buffer above.
   * @param[in] public_key A public key used as the certificate's subject key.
   * @param[in] public_key_size The size of the buffer above.
   * @param[in] inittime_custom_claims_buffer An optional flat buffer representing the init-time
   * custom claims.
   * @param[in] inittime_custom_claims_buffer_size The size of the buffer above in bytes.
   * @param[in] runtime_custom_claims_buffer An optional flat buffer representing the run-time
   * cutom claims.
   * @param[in] runtime_custom_claims_buffer_size The size of the buffer above in bytes.

>    * @param[in] attestation_results_buffer An optional flat buffer representing the attestation

Reviewer 4: Not sure how adding `attestatin_results_buffer` to this `oe_get_attestation_certificate_with_evidence_v3()` API is the best approach to add support of the passport mode. Does this API always include evidence in the output cert? But in passport mode, the attestation results is supposed to be carried instead of the evidence.

Author: You're correct, the example of passport mode below does not require evidence to be embedded. The APIs probably need different names.
